AK/String.cpp

```diff
@@ -239,9 +246,7 @@ Optional<size_t> String::find_byte_offset(StringView substring, size_t from_byte
 
 bool String::operator==(FlyString const& other) const
 {
-    if (reinterpret_cast<uintptr_t>(m_data) == other.data({}))
-        return true;
-    return bytes_as_string_view() == other.bytes_as_string_view();
+    return static_cast<StringBase>(*this) == other.data({});
```

Comment:
Not sure if this is actually correct, since this (a non-interned string) may have arbitrary string data, even if it's the same as *some* interned string data? As above you should instead intern this string and perform a FlyString comparison.

Reply:
Closed three other conversation in favor of one centralized place for discussion of `operator==` correctness and performance.

All comparison operators trivially decay to `StringBase::operator==(StringBase const&)` (they did so before but now I tried to make it especially obvious). The mentioned operator looks like this:
```cpp
if (is_short_string())
    return m_data == other.m_data;
if (other.is_short_string())
    return false;
if (m_data->is_fly_string() && other.m_data->is_fly_string())
    return m_data == other.m_data;
return bytes() == other.bytes();
```
where `m_data` is pointer to something (`Detail::StringData`) fully aliased with `ShortString`. Since short string can never be the same as `Detail::StringData` contents (there are several asserts checking that its length > `MAX_SHORT_STRING_LENGTH`), first two conditions are trivially correct and are $\mathcal{O}(1)$. Next, I check if both of the `StringData` classes have fly string bit set, and if they do, they are always deduplicated, so it is enough to check that pointers to them are equal. Note that this fully covers `FlyString::operator==(FlyString)` and `String::operator==(FlyString)` when String is FlyString in disguise. After that, I can't do anything other that do a slow Span comparison which old implementation did too.

Reply:
That .data (and the likes) confused me, since it wasn't clear to me which comparison actually gets called in the end. Okay.